src/NUnitEngine/nunit.engine.core/Services/ExtensionManager.cs

```diff
@@ -23,11 +24,24 @@ public class ExtensionManager : IExtensionManager
         //private readonly IAddinsFileReader _addinsReader;
         private readonly IDirectoryFinder _directoryFinder;
 
+        // List of all ExtensionPoints discovered
         private readonly List<ExtensionPoint> _extensionPoints = new List<ExtensionPoint>();
-        private readonly Dictionary<string, ExtensionPoint> _pathIndex = new Dictionary<string, ExtensionPoint>();
 
-        private readonly List<ExtensionNode> _extensions = new List<ExtensionNode>();
-        private readonly List<ExtensionAssembly> _assemblies = new List<ExtensionAssembly>();
+        // Index to ExtensionPoints based on the Path
+        private readonly Dictionary<string, ExtensionPoint> _extensionPointIndex = new Dictionary<string, ExtensionPoint>();
+
+        // List of ExtensionNodes for all extensions discovered.
+        private List<ExtensionNode> _extensions = new List<ExtensionNode>();
+
+        private bool _extensionsAreLoaded;
```

Comment:
The down side of initializing this here is that you don't get exceptions if you access it from a method not protected by `EnsureExtensionsAreLoaded`. But behave as if no extension are loaded.

Alternatively you could use `Lazy<T>` and then use `_extensions.Value`.

Reply:
Yes, it's a tradeoff. I ran tests for several days and I think all methods are now protected. If an error comes up in future due to this, I think the cause should be fairly clear. I'll leave this at least for now.